File: 2021/day02-1.ts
type Position = {
  forward: number;
  depth: number;
};

type Instruction = {
  direction: string;
  amount: number;
};

function solution(content: string) {
  const position = content
    .split("\n")
    .map((item) => parse(item))
    .reduce((acc, cur) => move(acc, cur), { depth: 0, forward: 0 });
  return position.depth * position.forward;
}

function move(position: Position, instruction: Instruction): Position {
  const { amount, direction } = instruction;
  const { depth, forward } = position;
  if (direction === "forward") {
    return { depth, forward: forward + amount };
  }
  if (direction === "up") {
    return { depth: depth - amount, forward };
  }
  if (direction === "down") {
    return { depth: depth + amount, forward };
  }
  return position;
}

function parse(line: string): Instruction {
  const split = line.split(" ");
  return { amount: Number(split[1]), direction: split[0] };
}

Deno.readTextFile("day02.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2021/day02_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any

type Position = tuple[int, int]
type Instruction = tuple[str, int]


def solution(content: list[str]) -> Any:
    instructions = [parse(line) for line in content]
    position = (0, 0)
    for instruction in instructions:
        position = move(position, instruction)
    return position[0] * position[1]


def move(position: Position, instruction: Instruction):
    direction, amount = instruction
    forward, depth = position
    if direction == "forward":
        return (forward + amount, depth)
    if direction == "up":
        return (forward, depth - amount)
    if direction == "down":
        return (forward, depth + amount)
    return position


def parse(line: str) -> Instruction:
    return (line.split(" ").pop(0), int(line.split(" ").pop()))


def main() -> None:
    with open("day02.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()
